File: narrow_road.py
import threading
import time
import random

class NarrowRoad:
	def __init__(self, max_cars):
		self.max_cars = max_cars
		self.current_cars = 0
		#self.road_lock = threading.Lock()
		self.left_gate = threading.Semaphore(1)
		self.right_gate = threading.Semaphore(0)
		self.left_cars = 20
		self.right_cars = 20

	def left_gatekeeper(self):
		while True:
			if self.left_gate.acquire():
				while self.current_cars < self.max_cars and self.left_cars > 0:
					self.current_cars += 1
					self.left_cars -= 1
					print(f"   Car from left gate entered. Current cars: {self.current_cars}")
					time.sleep(1)
				self.right_gate.release()
				self.current_cars = 0
				time.sleep(3)

	def right_gatekeeper(self):
		while True:
			if self.right_gate.acquire():
				while self.current_cars < self.max_cars and self.right_cars > 0:
					self.current_cars += 1
					self.right_cars -= 1
					print(f"   Car from right gate entered. Current cars: {self.current_cars}")
					time.sleep(1)
				self.left_gate.release()
				self.current_cars = 0
				time.sleep(3)

	def left_car(self):
		while True:
			time.sleep(random.randint(1, 10))
			self.left_cars += 1
			print("Car from left side arrived. Number of cars: ", self.left_cars)


	def right_car(self):
		while True:
			time.sleep(random.randint(1, 10))
			self.right_cars += 1
			print("Car from right side arrived. Number of cars: ", self.right_cars)


if __name__ == "__main__":
	road = NarrowRoad(5)
	threading.Thread(target=road.left_gatekeeper).start()
	threading.Thread(target=road.right_gatekeeper).start()
	threading.Thread(target=road.left_car).start()
	threading.Thread(target=road.right_car).start()

File: narrow_road.java
import java.util.concurrent.Semaphore;
import java.util.Random;

class NarrowRoad {
	private int maxCars;
	private int currentCars;
	private Semaphore leftGate;
	private Semaphore rightGate;
	private int leftCars;
	private int rightCars;
	private Random random;

	public NarrowRoad(int maxCars) {
		this.maxCars = maxCars;
		this.currentCars = 0;
		this.leftGate = new Semaphore(1);
		this.rightGate = new Semaphore(0);
		this.leftCars = 20;
		this.rightCars = 20;
		this.random = new Random();
	}

	public void leftGatekeeper() throws InterruptedException {
		while (true) {
			if (this.leftGate.tryAcquire()) {
				while (this.currentCars < this.maxCars && this.leftCars > 0) {
					this.currentCars++;
					this.leftCars--;
					System.out.println("   Car from left gate entered. Current cars: " + this.currentCars);
					Thread.sleep(1000);
				}
				if (this.currentCars == this.maxCars) 
					System.out.println("   Left gate allow maximum number of cars. Turning to right gate to open...");
				if (this.leftCars == 0) 
					System.out.println("   No more cars from left side. Turning to right gate to open...");

				this.rightGate.release();
				this.currentCars = 0;
				Thread.sleep(3000); 
			}
		}
	}

	public void rightGatekeeper() throws InterruptedException {
		while (true) {
			if (this.rightGate.tryAcquire()) {
				while (this.currentCars < this.maxCars && this.rightCars > 0) {
					this.currentCars++;
					this.rightCars--;
					System.out.println("   Car from right gate entered. Current cars: " + this.currentCars);
					Thread.sleep(1000);
				}
				if (this.currentCars == this.maxCars) 
					System.out.println("   Right gate allow maximum number of cars. Turning to left gate to open...");
				
				if (this.rightCars == 0) 
					System.out.println("   No more cars from right side. Turning to left gate to open...");
				
				this.leftGate.release();
				this.currentCars = 0;
				Thread.sleep(3000);
			}
		}
	}

	public void leftCar() throws InterruptedException {
		while (true) {
			Thread.sleep((random.nextInt(10) + 1)*1000);
			this.leftCars++;
			System.out.println("Car from left side arrived. Number of cars: " + this.leftCars);
		}
	}

	public void rightCar() throws InterruptedException {
		while (true) {
			Thread.sleep((random.nextInt(10) + 1)*1000);
			this.rightCars++;
			System.out.println("Car from right side arrived. Number of cars: " + this.rightCars);
		}
	}

	public static void main(String[] args) {
		NarrowRoad road = new NarrowRoad(5);
		new Thread(() -> {
			try {
				road.leftGatekeeper();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}).start();
		new Thread(() -> {
			try {
				road.rightGatekeeper();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}).start();
		new Thread(() -> {
			try {
				road.leftCar();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}).start();
		new Thread(() -> {
			try {
				road.rightCar();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}).start();
	}
}